File: mcp-sdk/exceptions.py
"""
Exception classes for Lattice MCP SDK

Custom exceptions for different error scenarios.
"""

from typing import Optional, Dict, Any


class LatticeError(Exception):
    """Base exception for all Lattice SDK errors"""
    
    def __init__(
        self, 
        message: str, 
        code: Optional[str] = None,
        context: Optional[Dict[str, Any]] = None
    ):
        super().__init__(message)
        self.message = message
        self.code = code or "LATTICE_ERROR"
        self.context = context or {}
    
    def __str__(self):
        return f"{self.code}: {self.message}"
    
    def to_dict(self) -> Dict[str, Any]:
        """Convert exception to dictionary"""
        return {
            "error": self.code,
            "message": self.message,
            "context": self.context
        }


class AuthenticationError(LatticeError):
    """Raised when authentication fails"""
    
    def __init__(self, message: str = "Authentication failed"):
        super().__init__(message, "AUTHENTICATION_ERROR")


class AuthorizationError(LatticeError):
    """Raised when user lacks required permissions"""
    
    def __init__(self, message: str = "Insufficient permissions"):
        super().__init__(message, "AUTHORIZATION_ERROR")


class ProjectNotFoundError(LatticeError):
    """Raised when a project is not found"""
    
    def __init__(self, project_id: Optional[str] = None):
        message = f"Project not found: {project_id}" if project_id else "Project not found"
        super().__init__(message, "PROJECT_NOT_FOUND")
        self.context = {"project_id": project_id} if project_id else {}


class MutationError(LatticeError):
    """Raised when a mutation operation fails"""
    
    def __init__(
        self, 
        message: str, 
        mutation_id: Optional[str] = None,
        stage: Optional[str] = None
    ):
        super().__init__(message, "MUTATION_ERROR")
        self.context = {
            "mutation_id": mutation_id,
            "stage": stage
        }


class ValidationError(LatticeError):
    """Raised when input validation fails"""
    
    def __init__(
        self, 
        message: str, 
        field: Optional[str] = None,
        value: Optional[Any] = None
    ):
        super().__init__(message, "VALIDATION_ERROR")
        self.context = {
            "field": field,
            "value": str(value) if value is not None else None
        }


class NetworkError(LatticeError):
    """Raised when network operations fail"""
    
    def __init__(self, message: str, status_code: Optional[int] = None):
        super().__init__(message, "NETWORK_ERROR")
        self.context = {"status_code": status_code} if status_code else {}


class RateLimitError(LatticeError):
    """Raised when rate limits are exceeded"""
    
    def __init__(
        self, 
        message: str = "Rate limit exceeded", 
        retry_after: Optional[int] = None
    ):
        super().__init__(message, "RATE_LIMIT_ERROR")
        self.context = {"retry_after": retry_after} if retry_after else {}


class ConfigurationError(LatticeError):
    """Raised when SDK configuration is invalid"""
    
    def __init__(self, message: str, setting: Optional[str] = None):
        super().__init__(message, "CONFIGURATION_ERROR")
        self.context = {"setting": setting} if setting else {}


class WebSocketError(LatticeError):
    """Raised when WebSocket operations fail"""
    
    def __init__(self, message: str, connection_id: Optional[str] = None):
        super().__init__(message, "WEBSOCKET_ERROR")
        self.context = {"connection_id": connection_id} if connection_id else {}


class TimeoutError(LatticeError):
    """Raised when operations timeout"""
    
    def __init__(self, message: str = "Operation timed out", timeout: Optional[int] = None):
        super().__init__(message, "TIMEOUT_ERROR")
        self.context = {"timeout": timeout} if timeout else {}


class OrganizationError(LatticeError):
    """Raised when organization operations fail"""
    
    def __init__(
        self, 
        message: str, 
        organization_id: Optional[str] = None
    ):
        super().__init__(message, "ORGANIZATION_ERROR")
        self.context = {"organization_id": organization_id} if organization_id else {}


class SubscriptionError(LatticeError):
    """Raised when subscription-related operations fail"""
    
    def __init__(
        self, 
        message: str, 
        subscription_id: Optional[str] = None,
        plan: Optional[str] = None
    ):
        super().__init__(message, "SUBSCRIPTION_ERROR")
        self.context = {
            "subscription_id": subscription_id,
            "plan": plan
        }


class APIKeyError(LatticeError):
    """Raised when API key operations fail"""
    
    def __init__(self, message: str, key_id: Optional[str] = None):
        super().__init__(message, "API_KEY_ERROR")
        self.context = {"key_id": key_id} if key_id else {}

File: mcp-sdk/nodejs-sdk/src/exceptions.ts
/**
 * Exception classes for Lattice MCP SDK
 * 
 * Custom error classes for different error scenarios.
 */

/**
 * Base error class for all Lattice SDK errors
 */
export class LatticeError extends Error {
  public readonly code: string;
  public readonly context: Record<string, any>;

  constructor(
    message: string,
    code: string = 'LATTICE_ERROR',
    context: Record<string, any> = {}
  ) {
    super(message);
    this.name = this.constructor.name;
    this.code = code;
    this.context = context;

    // Maintains proper stack trace for where our error was thrown (only available on V8)
    if (Error.captureStackTrace) {
      Error.captureStackTrace(this, this.constructor);
    }
  }

  toString(): string {
    return `${this.code}: ${this.message}`;
  }

  toJSON(): Record<string, any> {
    return {
      error: this.code,
      message: this.message,
      context: this.context
    };
  }
}

/**
 * Raised when authentication fails
 */
export class AuthenticationError extends LatticeError {
  constructor(message: string = 'Authentication failed') {
    super(message, 'AUTHENTICATION_ERROR');
  }
}

/**
 * Raised when user lacks required permissions
 */
export class AuthorizationError extends LatticeError {
  constructor(message: string = 'Insufficient permissions') {
    super(message, 'AUTHORIZATION_ERROR');
  }
}

/**
 * Raised when a project is not found
 */
export class ProjectNotFoundError extends LatticeError {
  constructor(projectId?: string) {
    const message = projectId ? `Project not found: ${projectId}` : 'Project not found';
    super(message, 'PROJECT_NOT_FOUND', projectId ? { project_id: projectId } : {});
  }
}

/**
 * Raised when a mutation operation fails
 */
export class MutationError extends LatticeError {
  constructor(
    message: string,
    mutationId?: string,
    stage?: string
  ) {
    super(message, 'MUTATION_ERROR', {
      mutation_id: mutationId,
      stage: stage
    });
  }
}

/**
 * Raised when input validation fails
 */
export class ValidationError extends LatticeError {
  constructor(
    message: string,
    field?: string,
    value?: any
  ) {
    super(message, 'VALIDATION_ERROR', {
      field: field,
      value: value !== undefined ? String(value) : undefined
    });
  }
}

/**
 * Raised when network operations fail
 */
export class NetworkError extends LatticeError {
  constructor(message: string, statusCode?: number) {
    super(message, 'NETWORK_ERROR', statusCode ? { status_code: statusCode } : {});
  }
}

/**
 * Raised when rate limits are exceeded
 */
export class RateLimitError extends LatticeError {
  constructor(
    message: string = 'Rate limit exceeded',
    retryAfter?: number
  ) {
    super(message, 'RATE_LIMIT_ERROR', retryAfter ? { retry_after: retryAfter } : {});
  }
}

/**
 * Raised when SDK configuration is invalid
 */
export class ConfigurationError extends LatticeError {
  constructor(message: string, setting?: string) {
    super(message, 'CONFIGURATION_ERROR', setting ? { setting: setting } : {});
  }
}

/**
 * Raised when WebSocket operations fail
 */
export class WebSocketError extends LatticeError {
  constructor(message: string, connectionId?: string) {
    super(message, 'WEBSOCKET_ERROR', connectionId ? { connection_id: connectionId } : {});
  }
}

/**
 * Raised when operations timeout
 */
export class TimeoutError extends LatticeError {
  constructor(message: string = 'Operation timed out', timeout?: number) {
    super(message, 'TIMEOUT_ERROR', timeout ? { timeout: timeout } : {});
  }
}

/**
 * Raised when organization operations fail
 */
export class OrganizationError extends LatticeError {
  constructor(message: string, organizationId?: string) {
    super(message, 'ORGANIZATION_ERROR', organizationId ? { organization_id: organizationId } : {});
  }
}

/**
 * Raised when subscription-related operations fail
 */
export class SubscriptionError extends LatticeError {
  constructor(
    message: string,
    subscriptionId?: string,
    plan?: string
  ) {
    super(message, 'SUBSCRIPTION_ERROR', {
      subscription_id: subscriptionId,
      plan: plan
    });
  }
}

/**
 * Raised when API key operations fail
 */
export class APIKeyError extends LatticeError {
  constructor(message: string, keyId?: string) {
    super(message, 'API_KEY_ERROR', keyId ? { key_id: keyId } : {});
  }
}

/**
 * Type guard to check if an error is a LatticeError
 */
export function isLatticeError(error: any): error is LatticeError {
  return error instanceof LatticeError;
}

/**
 * Helper function to create appropriate error from API response
 */
export function createErrorFromResponse(
  statusCode: number,
  responseData: any
): LatticeError {
  const message = responseData?.error || responseData?.message || 'Unknown error';
  const code = responseData?.code;

  switch (statusCode) {
    case 401:
      return new AuthenticationError(message);
    case 403:
      return new AuthorizationError(message);
    case 404:
      if (code === 'PROJECT_NOT_FOUND') {
        return new ProjectNotFoundError(responseData?.context?.project_id);
      }
      return new LatticeError(message, code || 'NOT_FOUND');
    case 422:
      return new ValidationError(
        message,
        responseData?.details?.[0]?.field,
        responseData?.details?.[0]?.context
      );
    case 429:
      return new RateLimitError(message, responseData?.retry_after);
    case 500:
    case 502:
    case 503:
    case 504:
      return new NetworkError(message, statusCode);
    default:
      return new LatticeError(message, code || 'UNKNOWN_ERROR', { status_code: statusCode });
  }
}